modules/imgproc/src/hal_replacement.hpp

```diff
@@ -1049,6 +1049,33 @@ inline int hal_ni_pyrdown(const uchar* src_data, size_t src_step, int src_width,
 #define cv_hal_pyrdown hal_ni_pyrdown
 //! @endcond
 
+/**
+   @brief Perform Gaussian Blur and downsampling for input tile with optional margins for submatrix
+   @param src_data Source image data
+   @param src_step Source image step
+   @param src_width Source image width
+   @param src_height Source image height
+   @param dst_data Destination image data
+   @param dst_step Destination image step
+   @param dst_width Destination image width
+   @param dst_height Destination image height
+   @param depth Depths of source and destination image
+   @param cn Number of channels
+   @param margin_left Left margins for source image
+   @param margin_top Top margins for source image
+   @param margin_right Right margins for source image
+   @param margin_bottom Bottom margins for source image
+   @param border_type Border type
+*/
+inline int hal_ni_pyrdown_offset(const uchar* src_data, size_t src_step, int src_width, int src_height,
```

Comment:
What is our plan to enable such feature in all filters. Starting with the name - shall we use _roi in the name? And pass the whole image size and roi inside?

How to make it python-compatible? (Numpy does not have a notion of roi)

Reply:
1. It's already done in current HAL API for filters.
2. Python and Numpy is not related here. Unfortunately we do not have solution for it.

Reply:
* Numpy supports array slicing (though I am not sure if the slices can be passed outside of the Python)
* Taking the ROI from source image then applying filter should produce the same result as doing it vice versa (i.e. applying filter then taking the ROI of the result).
  We can recommend Python users to stick to the second option at cost of performance.